File: [0188][Hard][Best_Time_to_Buy_and_Sell_Stock_IV]/Best_Time_to_Buy_and_Sell_Stock_IV.java
// solution 1, dp with general solution.
// leetcode time     cost : 5 ms
// leetcode memory   cost : 36.7 MB 
// Time  Complexity: O(kN)
// Space Complexity: O(k+N)

import java.util.Arrays;
class Solution {
    public int maxProfit(int max_k, int[] prices) {
        int n = prices.length;
        if (n == 0) 
            return 0;
        // 1 transaction consists of 1 buy and 1 sell, need at least 2 days, 
        // so for a effective constraint k_max, it must be smaller than n/2, otherwise k_max constraint will have no effect,
        // then it will be a infinity k_max, like problem 122.             
        if (max_k > n/2) 
            // or can  use K_max = (n//2) to update K_max to a effective value.        
            return maxProfit_k_inf(prices);        
        int dp[][][] = new int[n][max_k + 1][2];
        for (int i = 0; i < n; i++) {
            for (int k = max_k; k >= 1; k--) {
                if (i == 0) { 
                    dp[i][k][0]  = Math.max(0, Integer.MIN_VALUE + prices[i]) ;     //dp[i][k][0]  = 0;
                    dp[i][k][1]  = Math.max(Integer.MIN_VALUE, 0 - prices[i]) ;     //dp[i][k][0]  = - prices[0];
                    continue;
                }
                dp[i][k][0] = Math.max(dp[i-1][k][0], dp[i-1][k][1] + prices[i]);
                dp[i][k][1] = Math.max(dp[i-1][k][1], dp[i-1][k-1][0] - prices[i]);
            }
        }
        // enumerate n × max_k × 2 status.
        return dp[n - 1][max_k][0];
    }
    public int maxProfit_k_inf(int[] prices) {
        int maxprofit = 0;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > prices[i - 1])
                maxprofit += prices[i] - prices[i - 1];
        }
        return maxprofit;
    }
}

public class Best_Time_to_Buy_and_Sell_Stock_IV {
    public static void main(String args[]) {
        int k1 = 2,k2 = 4;
        int prices[] = {1,3,5,0,0,3,1,4}; // #expect is 8
        int prices_2[] = {3,2,6,5,0,3}; // #expect is 7
        Solution Solution_obj = new Solution();
        int result = Solution_obj.maxProfit(k1,prices);
        int result2 = Solution_obj.maxProfit(k2,prices_2);
        System.out.println("In java code, 1st result is :" + result + ", 2nd result is "+ result2);
    }
}

File: [0188][Hard][Best_Time_to_Buy_and_Sell_Stock_IV]/Best_Time_to_Buy_and_Sell_Stock_IV.py
# solution 1, dp general solution.
# leetcode time     cost : 212 ms
# leetcode memory   cost : 24.7 MB 
# Time  Complexity: O(N)
# Space Complexity: O(N)
import sys
class Solution:
    def maxProfit(self, K_max, prices) :
        if not prices:
            return 0
        n = len(prices)
        res = 0
        # 1 transaction consists of 1 buy and 1 sell, need at least 2 days, 
        # so for a effective constraint k_max, it must be smaller than n/2, otherwise k_max constraint will have no effect,
        # then it will be a infinity k_max, like problem 122.
        if (K_max > n//2):
            # also can use K_max = (n//2) to update K_max to a effective value.
            return self.maxProfit_k_inf(prices)
        profit = [[ [0 for _ in range(2)] for _ in range(K_max+1)] for _ in range(len(prices))] 

        # profit[i][k][j], i is the day, k is the transaction time, j is current holding status.
        # when k = any integer, i need to start from 0, initial value will be set in the loop k.
        for i in range(0,len(prices)):
            for k in range(1,K_max+1):
                if i == 0:
                   profit[i][k][0], profit[i][k][1] = 0, -prices[0] 
                   continue
                profit[i][k][1] = max(profit[i-1][k][1], profit[i-1][k-1][0] - prices[i]) 
                profit[i][k][0] = max(profit[i-1][k][0], profit[i-1][k][1] + prices[i]) 
     
        end = len(prices) - 1
        res = profit[end][K_max][0]
        return res 

    def maxProfit_k_inf(self, prices):
        print("maxProfit_k_inf is call")
        n = len(prices)
        dp_i_0 ,dp_i_1 =  0, -prices[0]
        for i in range(1,n):
            temp = dp_i_0
            dp_i_0 = max(dp_i_0, dp_i_1 + prices[i])
            dp_i_1 = max(dp_i_1, temp - prices[i])
        return dp_i_0; 

def main(): 
    k1 = 4   
    nums1 = [5,7,2,7,3,3,5,3,0] #expect is 9
    k2 = 2   
    nums1 = [3,2,6,5,0,3] #expect is 7    
    obj = Solution()
    result = obj.maxProfit(k1,nums1)
    print("return result is "+str(result));
    
if __name__ =='__main__':
    main() 